src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs

```diff
@@ -42,31 +42,7 @@ internal partial class HttpConnection : HttpConnectionBase, IDisposable
         private static readonly byte[] s_http1DotBytes = Encoding.ASCII.GetBytes("HTTP/1.");
         private static readonly ulong s_http10Bytes = BitConverter.ToUInt64(Encoding.ASCII.GetBytes("HTTP/1.0"));
         private static readonly ulong s_http11Bytes = BitConverter.ToUInt64(Encoding.ASCII.GetBytes("HTTP/1.1"));
-        private static readonly HashSet<KnownHeader> s_disallowedTrailers = new HashSet<KnownHeader>    // rfc7230 4.1.2.
-        {
-            // Message framing headers.
-            KnownHeaders.TransferEncoding, KnownHeaders.ContentLength,
-
-            // Routing headers.
-            KnownHeaders.Host,
-
-            // Request modifiers: controls and conditionals.
-            // rfc7231#section-5.1: Controls.
-            KnownHeaders.CacheControl, KnownHeaders.Expect, KnownHeaders.MaxForwards, KnownHeaders.Pragma, KnownHeaders.Range, KnownHeaders.TE,
-
-            // rfc7231#section-5.2: Conditionals.
-            KnownHeaders.IfMatch, KnownHeaders.IfNoneMatch, KnownHeaders.IfModifiedSince, KnownHeaders.IfUnmodifiedSince, KnownHeaders.IfRange,
-
-            // Authentication headers.
-            KnownHeaders.Authorization, KnownHeaders.SetCookie,
-
-            // Response control data.
-            // rfc7231#section-7.1: Control Data.
-            KnownHeaders.Age, KnownHeaders.Expires, KnownHeaders.Date, KnownHeaders.Location, KnownHeaders.RetryAfter, KnownHeaders.Vary, KnownHeaders.Warning,
-
-            // Content-Encoding, Content-Type, Content-Range, and Trailer itself.
-            KnownHeaders.ContentEncoding, KnownHeaders.ContentType, KnownHeaders.ContentRange, KnownHeaders.Trailer
-        };
+        private static readonly HashSet<KnownHeader> s_disallowedTrailers = new HashSet<KnownHeader>(KnownHeaders.TrailerDisallowedHeaders);
```

Comment:
Trailing headers are so rare that this probalby doesn't matter, but if we added a `HttpHeaderType.NonTrailing`, we could get rid of this `HashSet` all together. It is only used in one place:

https://github.com/dotnet/corefx/blob/3d724ba3038b4d439038364f898010e9e4189e21/src/System.Net.Http/src/System/Net/Http/SocketsHttpHandler/HttpConnection.cs#L902-L908

Reply:
I need a header name-based comparison in HttpResponseHeader because known headers might get converted to custom one which are compared by names. Thus, I cannot completely get rid of HashSet even if we introduce the new header type.

Reply:
You can use `KnownHeaders.TryGetKnownHeader` to do such a lookup already, so we'd be able to get rid of both hash sets.